File: core/exercises/03_http_post/solution/cpp/main.cpp
#include <httplib.h>
#include <nlohmann/json.hpp>
#include <iostream>
#include <string>

int main() {
    httplib::Client cli("httpbin.org");
    cli.set_connection_timeout(15, 0);
    nlohmann::json payload = {{"name", "codam"}};
    auto res = cli.Post("/post", payload.dump(), "application/json");
    if (!res || res->status != 200) {
        return 1;
    }
    auto data = nlohmann::json::parse(res->body);
    std::string name = data.at("json").at("name").get<std::string>();
    std::cout << "ECHO_OK:" << name << "\n";
    return 0;
}

File: core/exercises/03_http_post/solution/python/main.py
"""Exercise 03 — HTTP POST."""

import json
import os
import urllib.request


def main() -> None:
    url = os.environ.get("CODAMLINGS_ECHO_URL", "https://httpbin.org/post")
    payload = {"name": "codam"}
    body = json.dumps(payload).encode("utf-8")
    request = urllib.request.Request(url, data=body, method="POST")
    request.add_header("Content-Type", "application/json")

    with urllib.request.urlopen(request, timeout=15) as response:
        data = json.loads(response.read().decode("utf-8"))

    name = data["json"]["name"]
    print(f"ECHO_OK:{name}")


if __name__ == "__main__":
    main()
